### PR TITLE
not hardcode dns server address

### PR DESCRIPTION
**Pull request by joesis**

### CODE REVIEW

File: protected.go [hunk 1]
```diff
@@ -21,31 +21,36 @@ var (
...
+	dns     string
 }
 
-func New(protect Protect) *Protector {
-	return &Protector{protect}
+// New construct a protector from the protect function and DNS server address in host:port format.
+func New(protect Protect, dnsServerAddr string) (*Protector, error) {
+	host, port, err := SplitHostPort(dnsServerAddr)
+	if err != nil {
+		return nil, err
```
Comment (oxtoacart) [inline protected.go]: Maybe wrap the error so we have some context?

File: protected_test.go [hunk 2]
```diff
@@ -25,7 +25,7 @@ func (p *testprotector) Protect(fileDescriptor int) error {
 
 func TestConnectIP(t *testing.T) {
 	p := &testprotector{}
-	pt := New(p.Protect, "8.8.8.8")
+	pt, _ := New(p.Protect, "8.8.8.8:53")
```
Comment (oxtoacart) [inline protected_test.go]: Does it still work if we omit the DNS server port?

File: protected.go (hunk 1 above)
Comment (joesis) [inline protected.go]: fixed. I wrapped the error in `SplitHostPort` instead

File: protected_test.go (hunk 2 above)
Comment (joesis) [inline protected_test.go]: It will propagate the error return by `splitHostPort`  in that case

Comment (oxtoacart) [inline protected_test.go]: Is it defaulting the port to 53 if the caller doesn't supply it?

Comment (joesis) [inline protected_test.go]: Sorry for the late reply @oxtoacart. It won't default port to 53. I know it's a bit weird, but it just follows the Go convention and saves a few lines of code by reusing the same `splitHostPort` method.

Comment (oxtoacart) [inline protected_test.go]: Okay.

File: protected.go [hunk 3]
```diff
@@ -150,16 +145,16 @@ func (p *Protector) DialContext(ctx context.Context, network, addr string) (net.
 	chDone := make(chan bool)
 	go func() {
 		conn, err = p.dialContext(op, ctx, network, addr)
-		chDone <- true
-	}()
-	select {
-	case <-ctx.Done():
-		go func() {
-			<-chDone
+		select {
```
Comment (joesis) [inline protected.go]: This change avoids creating another goroutine, and saves one or two lines of code.

File: protected.go [hunk 4]
```diff
@@ -221,20 +206,15 @@ func (p *Protector) dialContext(op ops.Op, ctx context.Context, network, addr st
 	case <-ctx.Done():
 		return nil, ctx.Err()
 	default:
+		// continue
 	}
 
 	// Actually connect the underlying socket
-	err = conn.connectSocket()
+	err = conn.connectSocket(ctx)
```
Comment (joesis) [inline protected.go]: pass context down to `connectSocket` to remove the `select` below, and eliminate `connectTimeOut` altogether.

File: protected.go [hunk 5]
```diff
@@ -21,31 +21,35 @@ var (
...
 type Protector struct {
 	protect Protect
+	dnsAddr syscall.Sockaddr
+	dns     string
 }
 
-func New(protect Protect) *Protector {
-	return &Protector{protect}
+// New construct a protector from the protect function and DNS server address in host:port format.
+func New(protect Protect, dnsServerAddr string) (*Protector, error) {
```
Comment (joesis) [inline protected.go]: A few common conversion and error checking code are moved from `resolve` to here

File: protected.go [hunk 6]
```diff
@@ -321,18 +309,18 @@ func setQueryTimeouts(c net.Conn) {
...
-// SplitHostAndPort is a wrapper around net.SplitHostPort that also uses strconv
+// splitHostAndPort is a wrapper around net.SplitHostPort that also uses strconv
 // to convert the port to an int
-func SplitHostPort(addr string) (string, int, error) {
+func splitHostPort(addr string) (string, int, error) {
 	host, sPort, err := net.SplitHostPort(addr)
 	if err != nil {
 		log.Errorf("Could not split network address: %v", err)
-		return "", 0, err
+		return "", 0, errors.Wrap(err)
```
Comment (joesis) [inline protected.go]: Add stack trace

**Comment by joesis** [comment]
@oxtoacart reviewed this but in case @atavism you have some comments on this too.

File: protected.go [hunk 7]
```diff
@@ -21,31 +21,35 @@ var (
...
-	return &Protector{protect}
+// New construct a protector from the protect function and DNS server address in host:port format.
+func New(protect Protect, dnsServerAddr string) (*Protector, error) {
+	host, port, err := splitHostPort(dnsServerAddr)
+	if err != nil {
+		return nil, err
+	}
+	ipAddr := net.ParseIP(host)
+	if ipAddr == nil {
+		return nil, errors.New("invalid IP address")
```
Comment (oxtoacart) [inline protected.go]: Might be good to mention that it's an invalid DNS server ip address.

File: protected.go [hunk 8]
```diff
@@ -21,31 +21,35 @@ var (
...
+	dns     string
 }
 
-func New(protect Protect) *Protector {
-	return &Protector{protect}
+// New construct a protector from the protect function and DNS server address in host:port format.
+func New(protect Protect, dnsServerAddr string) (*Protector, error) {
+	host, port, err := splitHostPort(dnsServerAddr)
+	if err != nil {
+		return nil, err
```
Comment (oxtoacart) [inline protected.go]: Might be good to wrap with an error indicating that we had trouble parsing the DNS server address.

File: protected_test.go (hunk 2 above)
Comment (joesis) [inline protected_test.go]: @oxtoacart  I ended up passing dns IP instead, to bypass the IPv6 issue you encountered.

**Pull request merged**